framework/src/utils/SlepcSupport.C

```diff
@@ -920,6 +959,148 @@ PCSetUp_MoosePC(PC pc)
+  // as "doFreePowerIteration()" is false.
+  if (eigen_problem->doFreePowerIteration() && its == max_it && *reason <= 0)
+  {
+    *reason = EPS_CONVERGED_USER;
+    eps->nconv = 1;
+  }
+  return 0;
+}
+
+PetscErrorCode
+mooseSlepcEPSGetSNES(EPS eps, SNES * snes)
+{
+  PetscErrorCode ierr;
+  PetscBool same, nonlinear;
+
+  ierr = PetscObjectTypeCompare((PetscObject)eps, EPSPOWER, &same);
+  LIBMESH_CHKERR(ierr);
+
+  if (!same)
+    mooseError("It is not eps power, and there is no snes");
+
+  ierr = EPSPowerGetNonlinear(eps, &nonlinear);
+  LIBMESH_CHKERR(ierr);
+
+  if (!nonlinear)
+    mooseError("It is not a nonlinear eigen solver");
+
+  ierr = EPSPowerGetSNES(eps, snes);
+  LIBMESH_CHKERR(ierr);
+
+  return 0;
+}
+
+PetscErrorCode
+mooseSlepcEPSSNESSetUpOptionPrefix(EPS eps)
+{
+  PetscErrorCode ierr;
+  SNES snes;
+  const char * prefix = nullptr;
+
+  ierr = mooseSlepcEPSGetSNES(eps, &snes);
+  LIBMESH_CHKERR(ierr);
+  // There is an extra "eps_power" in snes that users do not like it.
+  // Let us remove that from snes.
+  // Retrieve option prefix from EPS
+  ierr = PetscObjectGetOptionsPrefix((PetscObject)eps, &prefix);
+  LIBMESH_CHKERR(ierr);
+  // Set option prefix to SNES
+  ierr = SNESSetOptionsPrefix(snes, prefix);
+  LIBMESH_CHKERR(ierr);
+
+  return 0;
+}
+
+PetscErrorCode
+mooseSlepcEPSSNESSetCustomizePC(EPS eps)
+{
+  PetscErrorCode ierr;
+  SNES snes;
+  KSP ksp;
+  PC pc;
+
+  // Get SNES from EPS
+  ierr = mooseSlepcEPSGetSNES(eps, &snes);
+  LIBMESH_CHKERR(ierr);
+  // Get KSP from SNES
+  ierr = SNESGetKSP(snes, &ksp);
+  LIBMESH_CHKERR(ierr);
+  // Get PC from KSP
+  ierr = KSPGetPC(ksp, &pc);
+  LIBMESH_CHKERR(ierr);
+  // Set PC type
+  ierr = PCSetType(pc, "moosepc");
+  LIBMESH_CHKERR(ierr);
+  return 0;
+}
+
+PetscErrorCode
+mooseSlepcEPSSNESKSPSetPCSide(FEProblemBase & problem, EPS eps)
+{
+  PetscErrorCode ierr;
+  SNES snes;
+  KSP ksp;
+
+  // Get SNES from EPS
+  ierr = mooseSlepcEPSGetSNES(eps, &snes);
+  LIBMESH_CHKERR(ierr);
+  // Get KSP from SNES
+  ierr = SNESGetKSP(snes, &ksp);
+  LIBMESH_CHKERR(ierr);
+
+  Moose::PetscSupport::petscSetDefaultPCSide(problem, ksp);
+
+  Moose::PetscSupport::petscSetDefaultKSPNormType(problem, ksp);
+  return 0;
+}
+
+PetscErrorCode
+mooseSlepcEPSMonitor(EPS /*eps*/,
+                     int its,
+                     int /*nconv*/,
+                     PetscScalar * eigr,
+                     PetscScalar * /*eigi*/,
+                     PetscReal * /*errest*/,
+                     int /*nest*/,
+                     void * mctx)
+{
+
+  EigenProblem * eigen_problem = static_cast<EigenProblem *>(mctx);
+  auto & console = eigen_problem->console();
+
+  auto inverse = eigen_problem->outputInverseEigenvalue();
+
+  auto eigenvalue = inverse ? 1.0 / (*eigr) : (*eigr);
+
+  // The term "k-eigenvalue" is adopted from the neutronics community.
+  console << " Iteration " << its << std::setprecision(8)
```

Comment:
Suggest to have precision 10 here. 8 is slightly on the margin i.e. nl_rel_tol may still not reached but eigenvalue is accurate in 8 precision.

Reply:
OK